### PR TITLE
Ensure vpubs not both nonzero

### LINKED ISSUE #466: Ensure that  vpub_in and vpub_out are not both nonzero
Pour A values:

```
vpub_in = 5
oldcoin_0 = 0
oldcoin_1 = 0
newcoin_0 = 1
newcoin_1 = 0
vpub_out = 4
```

Pour B values:

```
vpub_in = 1
oldcoin_0 = 0
oldcoin_1 = 0
newcoin_0 = 1
newcoin_1 = 0
vpub_out = 0
```

Do these pours have the same semantics? They both validate and both affect the "value pool" in the exact same way. One of them just requires the value pool has more value in it than is really necessary.

Perhaps this should be a well-formedness check on pours if it has anonymity implications? If not, we can add this constraint later in a softfork.

### PR DESCRIPTION
**Pull request by defuse**
(By @str4d and @defuse in our pairing)

Edit by sean: this closes #466

### CODE REVIEW

**Comment by defuse** [comment]
NACK: Needs tests.

**Comment by defuse** [comment]
(I'll write the tests.)

**Comment by ebfull** [comment]
Nice!

The RPC tests will also break from this, you'll want to fix:

https://github.com/zcash/zcash/blob/zc.v0.11.2.latest/qa/rpc-tests/zcpour.py#L27

and 

https://github.com/zcash/zcash/blob/zc.v0.11.2.latest/qa/rpc-tests/zcpourdoublespend.py#L59

That should be it.

File: src/main.cpp [hunk 1]
```diff
@@ -881,26 +881,36 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state)
...
                              REJECT_INVALID, "bad-txns-vpub_old-toolarge");
+        }
 
-        if (pour.vpub_new > MAX_MONEY)
+        if (pour.vpub_new > MAX_MONEY) {
             return state.DoS(100, error("CheckTransaction(): pour.vpub_new too high"),
                              REJECT_INVALID, "bad-txns-vpub_new-toolarge");
+        }
+
+        if (pour.vpub_new =! 0 && pour.vpub_old != 0) {
```
Comment (daira) [inline src/main.cpp]: Typo: ``s/=!/!=/``

(The current code is equivalent to ``pour.vpub_new = 1; if (pour.vpub_old != 0) { ...``. I'm surprised it didn't give a compiler warning.)

**Comment by daira** [comment]
NACK due to https://github.com/zcash/zcash/pull/874/files#r59992180

File: src/main.cpp (hunk 1 above)
Comment (defuse) [inline src/main.cpp]: Wow! (It may have given a warning, but the mass of warnings the build already spits out has trained me to ignore them.)

Comment (daira) [inline src/main.cpp]: Yeah, this is why we should fix the false-positive warnings.

**Comment by defuse** [comment]
@zkbot try

**Comment by defuse** [comment]
(I'm debugging buildbot at the moment)

**Comment by defuse** [comment]
@zkbot try

**Comment by ebfull** [comment]
@zkbot force

**Comment by defuse** [comment]
Superseded by #895

**Pull request closed without merging**